csrc/adagrad/cpu_adagrad.cpp

```diff
@@ -21,26 +21,19 @@ static std::unordered_map<int, std::shared_ptr<void>> s_optimizers;
 
 // C++ interface
 
-void Adagrad_Optimizer::Step_1(float* _params,
-                               float* grads,
+template <typename T, typename ds_device_precision_t>
+void Adagrad_Optimizer::Step_1(T* _params,
+                               T* grads,
                                float* _exp_avg_sq,
```

Comment:
This for consistency. 
```suggestion
                               T* _exp_avg_sq,
```

Reply:
Thanks for catching this.
Since fp32_optimizer_states may be controlled from configuration independently of params and grads dtype I think I need to add a third template argument T2 for state['exp_avg_sq']. Now the template will be controlled by 3 types, I think I need to find a way to call it at runtime without checking for each tensor type at runtime in ds_adagrad_step.
Also, the code in ds_adagrad_step_plus_copy is even more complicated and not even used since this path is disabled in stage_1_and_2.py::_optimizer_step()

Reply:
These are all the possible type 
```
params+grads, opt_states, fp16_params
float       , float     , half
float       , float     , bfloat
half        , half      , half
bfloat      , bfloat    , bfloat
half        , float     , half
bfloat      , float     , bfloat
```
Since the number is limited, I think I will expand the if statements in ds_adagrad_step (and matching functions for adam and lion) in order to call the correct template.
Is it ok?

Reply:
The requirements are actually simpler than you think for the following reasons. 
1. `params`, `grads`, and `opt_states` must have the same dtype. 
2. Clarification from @RezaYazdaniAminabadi confirms that template usage should eliminate `half_precision` argument and associated logic. 
3. As you observed, `ds_adagrad_step_plus_copy` is not used, as we were never able to realize the potential perf benefits. I am happy to deprecate this API if that makes things easier for you. 

So, the requirements can be expressed in terms of host vs device states:
1. host-side states = `params`, `grads`, and `opt_states`
2. device-side state = `fp16_params`

One caveat is that the above is true only for mixed-precision training, since DS also supports fp32 training. 

Furthermore, the host-side states are of the same type, i.e., `ds_host_precision_t`, while the device state is of `ds_device_precision_t`. These two typenames can be instantiated as follows (for mixed precision training):

```python
  params+grads+opt_states,      fp16_params
  float                                half
  float                                bfloat
  half                                 half
  bfloat                             bfloat                                         
```

Please let me know your thoughts and whether the above is helpful. 

@RezaYazdaniAminabadi, please share your thoughts as well.

Reply:
All good here, thanks @tjruwase for the clarification. Just wanna confirm that even though the optimizer states are sometimes saved in bit16 precision, we still wanna do the computation in fp32. So, maybe we can have a template argument as `compute_type`.

Reply:
@tjruwase 
Regarding 1, I see there is an option to set cpu_offload=True, fp_32_optimizer_states=True and fp16_master_weights_And_gradients=True. In this mode params+grads will be half and opt_states will be float. A different template type is required for each.

Regarding 3, if we deprecate ds_adagrad_step_plus_copy we don't need ds_device_precision_t anymore, since device_params is always passed as nullptr. Can we also remove all related code under __ENABLE_CUDA__ and __ENABLE_CANN__ ?

Reply:
> Regarding 1, I see there is an option to set cpu_offload=True, fp32_optimizer_states=True and fp16_master_weights_And_gradients=True. In this mode params+grads will be half and opt_states will be float. A different template type is required for each.

@BacharL, I was unaware of this option, even though @RezaYazdaniAminabadi alluded to it above. My search suggests it was introduced in the MoE PR #1310. In my opinion, we should deprecate this option as the costs outweigh the benefits. I would like to be convinced otherwise. I will now discuss the cost/benefit of this option as I see it. 

**Benefit**
The main benefit is reducing per-parameter optimizer footprint in CPU memory from 16 bytes to 12 bytes, i.e., 25% savings. There are also DtoH/HtoD b/w reductions, but I think those can also be realized in the baseline mode. Before discussing the downsides below, @RezaYazdaniAminabadi, @jeffra, and @awan-10 can you please share your thoughts on other benefits.

**Costs**
1. Use case is narrow: mixed-precision + offloading. Also, CPU memory savings is rarely critical compared to GPU memory.
2. Increased code complexity since the implementation has two manage 3 data types instead of 2 or 1 (if we further deprecate the `*_step_plus_copy()` API.
3. It complicates integration of other cpu-based optimizer implementations, such as torch Adam (works on both CPU and GPU). Ideally, CPUAdam, torch.Adam, (and others) should be drop-in replacements differing only in implementation details such as performance. However, introduction of [fp32_optimizer_states](https://github.com/microsoft/DeepSpeed/blob/0b224edcf7d83713b95ad6b989694a8bdf01809e/deepspeed/ops/adam/cpu_adam.py#L25) to support this option has deviated from torch API. 
4. It complicates our plans to support optimizers of arbitrary data types, besides fp32, such as 4-bit optimizers requested [here](https://github.com/microsoft/DeepSpeed/issues/5276). The problem is our deviation from the torch [approach](https://github.com/pytorch/pytorch/blob/82b7b59d2acd849cb16f2167fa3eaf7cabf95dcf/torch/optim/adam.py#L153-L159) of inferring optimizer state type from param type. 
5. It complicates model checkpointing, since [self.single_partition_of_fp32_groups](https://github.com/microsoft/DeepSpeed/blob/0b224edcf7d83713b95ad6b989694a8bdf01809e/deepspeed/runtime/zero/stage_1_and_2.py#L389-L402) which points to the master weights is widely used in checkpoint [save/load](https://github.com/microsoft/DeepSpeed/blob/0b224edcf7d83713b95ad6b989694a8bdf01809e/deepspeed/runtime/zero/stage_1_and_2.py#L2399). Right now, I am not sure how this option works with model checkpointing. 

Apologies for the very long post, I think it is important that we get this right. Thanks!

Reply:
> Regarding 3, if we deprecate ds_adagrad_step_plus_copy we don't need ds_device_precision_t anymore since device_params is always passed as nullptr. Can we also remove all related code under `__ENABLE_CUDA__` and `__ENABLE_CANN__` ?

Yes, I generally love code removal whenever possible.

Reply:
> > Regarding 3, if we deprecate ds_adagrad_step_plus_copy we don't need ds_device_precision_t anymore since device_params is always passed as nullptr. Can we also remove all related code under `__ENABLE_CUDA__` and `__ENABLE_CANN__` ?
> 
> Yes, I generally love code removal whenever possible.

Cleaned up all device code from cpu optimizers
Now left with 2 template types, params+grads and optimizer state.

Reply:
> Now left with 2 template types, params+grads and optimizer state.

Thanks @BacharL for the updates. Please note that if I get alignment on my proposal regarding 1, then we would be left with one template type for params+grads+optimizer state. However, I think that could be a separate PR, so as not to delay this any further. 

One request that I do have is for performance benchmarking of the new code to uncover any regressions. The code [here](https://github.com/microsoft/DeepSpeed/tree/master/tests/perf) might be useful starting point.

Reply:
Thanks for unblocking this PR

Benchmarked Adam and Adagrad using Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz
CPUAdam fp32  0.400s -> 0.399s
CPUAdam fp16  0.290s -> 0.283s
CPUAdam bf16  NA      -> 0.298s

CPUAdagrad fp32  0.291s -> 0.298s
CPUAdagrad fp16  NA      -> 0.180s
CPUAdagrad bf16  NA      -> 0.191s

Reply:
@BacharL, thanks for sharing these results. Did you use the existing perf script? If not, do you mind sharing yours?

Reply:
Yes, I used the perf script above. just changed param dtype to the desired one.